keras_nlp/applications/bert.py

```diff
@@ -0,0 +1,372 @@
+        )
+        self.loss_tracker = keras.metrics.Mean(name="loss")
+        self.lm_loss_tracker = keras.metrics.Mean(name="lm_loss")
+        self.nsp_loss_tracker = keras.metrics.Mean(name="nsp_loss")
+        self.lm_accuracy = keras.metrics.SparseCategoricalAccuracy(
+            name="lm_accuracy"
+        )
+        self.nsp_accuracy = keras.metrics.SparseCategoricalAccuracy(
+            name="nsp_accuracy"
+        )
+
+    def call(self, data):
+        sequence_output, pooled_output = self.encoder(
+            {
+                "input_ids": data["input_ids"],
+                "input_mask": data["input_mask"],
+                "segment_ids": data["segment_ids"],
+            }
+        )
+        lm_preds = self.masked_lm_head(
+            sequence_output, data["masked_lm_positions"]
+        )
+        nsp_preds = self.next_sentence_head(pooled_output)
+        return lm_preds, nsp_preds
+
+    def train_step(self, data):
+        with tf.GradientTape() as tape:
+            lm_preds, nsp_preds = self(data, training=True)
+            lm_labels = data["masked_lm_ids"]
+            lm_weights = data["masked_lm_weights"]
+            nsp_labels = data["next_sentence_labels"]
+
+            lm_loss = keras.losses.sparse_categorical_crossentropy(
+                lm_labels, lm_preds, from_logits=True
+            )
+            lm_weights_summed = tf.reduce_sum(lm_weights, -1)
+            lm_loss = tf.reduce_sum(lm_loss * lm_weights, -1)
+            lm_loss = tf.math.divide_no_nan(lm_loss, lm_weights_summed)
+            nsp_loss = keras.losses.sparse_categorical_crossentropy(
+                nsp_labels, nsp_preds, from_logits=True
+            )
+            nsp_loss = tf.reduce_mean(nsp_loss)
+            loss = lm_loss + nsp_loss
+
+        # Compute gradients
+        trainable_vars = self.trainable_variables
+        gradients = tape.gradient(loss, trainable_vars)
+        # Update weights
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+
+        # Update metrics
+        self.loss_tracker.update_state(loss)
+        self.lm_loss_tracker.update_state(lm_loss)
+        self.nsp_loss_tracker.update_state(nsp_loss)
+        self.lm_accuracy.update_state(lm_labels, lm_preds, lm_weights)
+        self.nsp_accuracy.update_state(nsp_labels, nsp_preds)
+        return {m.name: m.result() for m in self.metrics}
+
+
+class BertClassifier(keras.Model):
+    """Classifier model with BertEncoder."""
+
+    def __init__(self, encoder, num_classes, **kwargs):
+        super().__init__(**kwargs)
+        self.encoder = encoder
+        self.num_classes = num_classes
+        self._logit_layer = keras.layers.Dense(
+            num_classes,
+            kernel_initializer=encoder.initializer_fn,
+            name="logits",
+        )
+
+    def call(self, inputs):
+        # Ignore the sequence output, use the pooled output.
+        _, pooled_output = self.bert_model(inputs)
```

Comment:
Does this actually work? You are returning a dict above and unpacking it as a tuple.

Reply:
Not tested. Will remove with `BertLanguageModel`

Reply:
Good call. Fixed.